### PR TITLE
Rename nanopb::String to nanopb::ByteString

### PR DESCRIPTION
**Pull request by wilhuff**
... and flesh it out.

This prepares for making `ByteString` the type that represents blobs and other things like resume tokens. The name intentionally mirrors `ByteString` from Android libprotobuf.

### CODE REVIEW

File: Firestore/core/test/firebase/firestore/remote/datastore_test.mm [hunk 1]
```diff
@@ -105,7 +105,7 @@ void CancelLastCall() {
       : worker_queue{std::make_shared<AsyncQueue>(
             absl::make_unique<ExecutorLibdispatch>(dispatch_queue_create(
                 "datastore_test", DISPATCH_QUEUE_SERIAL)))},
-        database_info{DatabaseId{"p", "d"}, "", "some.host", false},
+        database_info{DatabaseId{"p", "d"}, "", "localhost", false},
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/remote/datastore_test.mm]: No objection, though this seems unrelated.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  return MakeBytesArray(reinterpret_cast<const uint8_t*>(data), size);
+}
+
+}  // namespace
+
+ByteString::ByteString(const std::vector<uint8_t>& value)
+    : bytes_(MakeBytesArray(value.data(), value.size())) {
+}
+
+ByteString::ByteString(const pb_bytes_array_t* bytes)
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: This ctor isn't declared in the header? I don't think you need it.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc [hunk 3]
```diff
@@ -14,65 +14,65 @@
...
+TEST(ByteStringTest, DefaultConstructor) {
+  ByteString str;
   EXPECT_EQ(nullptr, str.data());
 }
 
-TEST(String, FromStdString) {
+TEST(ByteStringTest, FromStdString) {
   std::string original{"foo"};
-  String copy{original};
+  ByteString copy{original};
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: std::string isn't implicitly convertible to ByteString, so I don't see how the op== on line 33 could work. (nanopb::String worked because it's op== allowed the second param to be an string_view.)

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: Sorry, this bled through from a later commit state where it is needed. Removed.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ByteString::ByteString(const ByteString& other)
+    : bytes_{MakeBytesArray(other.data(), other.size())} {
+}
+
+ByteString::ByteString(ByteString&& other) noexcept : ByteString{} {
+  swap(*this, other);
+}
+
+ByteString::~ByteString() {
+  std::free(bytes_);
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: Done.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2018 Google
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: I was torn about this. It's a rename and mild modification of an artifact released in 2018. Copyright is on the contents not the filename. If you object to this line of thinking I'll change it, but this wasn't an oversight.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: So I'll admit it: I tried to extract this from a much larger change and failed to get it all in here. The test should really be about testing equality of `vector<uint8_t>`. I've pulled those tests forward.

Unfortunately this adds a new method to `ByteString` that I was hoping to leave out of this review: `CopyVector`. I'm not sure I like that name in preference to something simpler like `ToVector`, but I also don't want it to seem as if that vector is cheap. Anyway, PTAL and LMK.

File: Firestore/core/test/firebase/firestore/remote/datastore_test.mm (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/datastore_test.mm]: Sorry this is unrelated but it was required to make the tests work on a plane :-D

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: OK. Renamed.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 6]
```diff
@@ -95,6 +95,10 @@ pb_bytes_array_t* ByteString::release() {
   return result;
 }
 
+std::vector<uint8_t> ByteString::CopyVector() const {
+  return std::vector<uint8_t>{bytes_->bytes, bytes_->bytes + bytes_->size};
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: Done.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc [hunk 7]
```diff
@@ -16,12 +16,23 @@
 
 #include "Firestore/core/src/firebase/firestore/nanopb/byte_string.h"
 
+#include <cstdlib>
+
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: Done.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc [hunk 8]
```diff
@@ -16,12 +16,23 @@
...
 #include "gtest/gtest.h"
 
 namespace firebase {
 namespace firestore {
 namespace nanopb {
 
+namespace {
+
+std::vector<uint8_t> ToVector(const std::string& str) {
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: Resolved by making it `ByteString::ToVector`.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc [hunk 9]
```diff
@@ -30,19 +41,19 @@ TEST(ByteStringTest, DefaultConstructor) {
 TEST(ByteStringTest, FromStdString) {
   std::string original{"foo"};
   ByteString copy{original};
-  EXPECT_EQ(copy, original);
+  EXPECT_EQ(copy.CopyVector(), ToVector(original));
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: Done.
